### PR TITLE
Add missing perms (Py 3, XFCE4, browser, os, ...)

### PR DESCRIPTION
**Pull request by LorenzoAncora**
Useful additions:
- OS: implicit ALLOW directives to accept read-only access to informative system files like `os-release` and `machine-id`;
- Fonts: extended pattern matching;
- NetworkManager and similar: implicit ALLOW directives to accept read-only access to DNS parameters;
- XFCE4, snapd: implicit ALLOW directives to accept read-only access to the application list and to the MIME cache;
- Firefox ESR: implicit ALLOW directives to allow direct execution (AA profile for Firefox already exists) from the About dialog;
- Python 3 plugin: implicit ALLOW directives to allow the enumeration of available modules.

<sub>Note: this commit was designed with GNU/Linux Debian in mind.</sub>

### CODE REVIEW

File: etc/apparmor.d/abstractions/xchat-based
Comment (adrelanos) [inline etc/apparmor.d/abstractions/xchat-based:62]: > U - exec to unconfined, with environment variable scrubbing

Not safe.

> P - exec to profile that matches executable name, with environment variable scrubbing

Sounds better.

Could you change that and test please?

File: etc/apparmor.d/abstractions/xchat-based
Comment (adrelanos) [inline etc/apparmor.d/abstractions/xchat-based:28]: Is write access really required? Did you have a denied message without write access?

**Review by adrelanos — CHANGES_REQUESTED** [review]
Looks good overall. Some changes requested.

**Comment by adrelanos** [comment]
Is the hexchat -> about link really worth allowing firefox to be executed from xchat?

How do to other profiles handle such interactions?

//cc @HulaHoopWhonix @Algernon-01 @troubadoour

File: etc/apparmor.d/abstractions/xchat-based
Comment (LorenzoAncora) [reply etc/apparmor.d/abstractions/xchat-based:28]: > Is write access really required?

@adrelanos  denying access to this special device file will compromise the present or future stability of the application: POSIX.1-2017 _requires_ `/dev/null` to be _always_ readable and writable by all applications and their subcomponents (modules, extensions, etc.).

File: etc/apparmor.d/abstractions/xchat-based
Comment (LorenzoAncora) [reply etc/apparmor.d/abstractions/xchat-based:62]: > Not safe.

@adrelanos an AppArmor profile for Mozilla Firefox is already available and will ensure a safe execution when enforced. However, it must be customized by the user based on the extensions used.
In this case using `Ux` is not a great weakness, because Mozilla Firefox itself already uses a sandbox.

> Could you change that and test please?

Good idea but I've already tested this combination on multiple computers: without `Ux` Mozilla Firefox won't start ("profile transition not found"). If the Firefox profile is present and enabled, it will be imposed no matter what.
I believe that `Ux` is a reasonable compromise and that, given the particular architecture of Firefox - unlike `ux` - does not compromise security.

**Comment by LorenzoAncora** [comment]
> Is the hexchat -> about link really worth allowing firefox to be executed from xchat?

@adrelanos it is not up to us to decide, the profile must guarantee the full functionality of the application. The author's website contains documentation and references: support links must always work, because the user using HexChat may not be the administrator who has access to the AppArmor configuration.

**Comment by adrelanos** [comment]
Shouldn't that be part of apparmor abstractions/base then?

**Comment by LorenzoAncora** [comment]
> Shouldn't that be part of apparmor abstractions/base then?

That's why I can safely tell you that access is necessary, in fact it is already part of the basic abstractions:

```
  # /dev/null is pretty harmless and frequently used
  /dev/null                      rw,
  # as is /dev/zero
  /dev/zero                      rw,
```
I have included it because access to `/dev/null` is _critical_ to interpreters running some of the HexChat plugins. Since such extensions can be used to interact with databases using shell commands (a common alternative to direct library access), if one of these commands fails due to an unusable device file, the database would remain in an inconsistent state or cause uncontrolled accumulation of data. A typical example is to use PostgreSQL as a logging backend: if the `write` access right to `/dev/null` is lost because the administrator has customized/removed `abstractions/base` the failure of some of these commands chains could lead to non-compaction, causing in the best case a progressive slowdown of HexChat and all other applications that use the database.
Ensuring access to `/dev/null` from the interpreters is imperative.

**Comment by LorenzoAncora** [comment]
> Shouldn't that be part of apparmor abstractions/base then?

Otherwise, if your question was about Firefox, `abstractions/base` is reserved for the common needs of all programs. It would be wiser to avoid modifying it for a single application.

**Comment by adrelanos** [comment]
Lorenzo Ancora:
>> Shouldn't that be part of apparmor abstractions/base then?
> 
> Otherwise, if your question was about Firefox, `abstractions/base` is reserved for the common needs of all programs. It would be wiser to avoid modifying it for a single application.
> 

No, I didn't suggest modifying abstractions/base. (If we were to do
that, such a patch should be submitted to upstream.)

**Comment by adrelanos** [comment]
Since "/dev/null rw," is already part of abstractions/base and since we
do import abstractions/base in this apparmor profile already, it should
be removed here to avoid duplication and keep the profile shorter.

> because the administrator has customized/removed `abstractions/base`

That is entirely unsupported. There's a ton of ways on linux where
administrators can modify files which can have bad consequences. Those
who modify `abstractions/base` should know what they are doing. Trying
to foresee what kind of bad changes users can make is futile. It's
simply too much. So better not get started with it. Otherwise we end up
duplicating too many things.

I am willing to reconsider if other apparmor profiles (already in Debian
or other upstream resource) are doing this and/or if this is some kind
of convention of doing so.

**Comment by adrelanos** [comment]
Btw would you be interested to submit this profile upstream to Debian and/or HexChat?

**Comment by LorenzoAncora** [comment]
> That is entirely unsupported. 

Don't worry, if this were true, such rules would be hard-coded or at least there would be a comment in the beginning of `abstractions/base` with a warning. :-)
To further assure you that administrators can edit `abstractions/base`, the file begins with this comment:

```
  # (Note that the ldd profile has inlined this file; if you make
  # modifications here, please consider including them in the ldd
  # profile as well.)
```

This implies that administrators can freely alter this configuration file to satisfy their security requisites.

>  I am willing to reconsider if other apparmor profiles (already in Debian or other upstream resource) are doing this and/or if this is some kind of convention of doing so.

It is a question of wisdom to know how to impose rules at critical points, but in this case the stability of the application certainly depends on these measures.
For example the default Debian profile of a complex application like LibreOffice imposes this rule because it manages personal/sensitive user data using one or more databases.

> Btw would you be interested to submit this profile upstream to Debian and/or HexChat?

Of course! :-)
As I understand this repository is the main source where to find this profile. However, after you have done the merge here I will ask the package maintainers if they can distribute the most recent version of the profile, although in the case of Debian this can be a slow process.

File: etc/apparmor.d/abstractions/xchat-based
Comment (adrelanos) [reply etc/apparmor.d/abstractions/xchat-based:62]: Debian package apparmor-profiles-extra contains /etc/apparmor.d/usr.bin.pidgin which then uses:

    #include <abstractions/ubuntu-browsers>

(Also usr.bin.evince and usr.bin.thunderbird is using it.)

Looks better / more following convention / supporting more browsers. What about that one? Could you try please?

(abstractions/ubuntu-browsers is shipped by Debian apparmor package.)

Comment (LorenzoAncora) [reply etc/apparmor.d/abstractions/xchat-based:62]: > Debian package apparmor-profiles-extra contains /etc/apparmor.d/usr.bin.pidgin which then uses:
> #include <abstractions/ubuntu-browsers>

Yes, this seemed like a good idea in the beginning but then I've discovered some important details that discouraged me from using the abstraction.
The initial comments inside `abstractions/ubuntu-browsers` says:
```
abstraction for allowing access to graphical browsers in Ubuntu
```
Probably it is called `ubuntu-browsers` because it only guarantees long-term compatibility with Ubuntu and derivatives. However, Debian and many other distributions do not descend from Ubuntu. The incompatibility is also demonstrated by:

```
# Users of this abstraction need to #include the ubuntu-helpers abstraction
# in the toplevel profile. Eg:
# #include <abstractions/ubuntu-helpers>
```
Here the word Ubuntu is repeated like some sort of warning.
If we look inside `ubuntu-helpers` there is a very important comment:
```
# Lenient profile that is intended to be used when 'Ux' is desired but
# does not provide enough environment sanitizing. This effectively is an
# open profile that blacklists certain known dangerous files and also
# does not allow any capabilities. For example, it will not allow 'm' on files
# owned be the user invoking the program. While this provides some additional
# protection, please use with care as applications running under this profile
# are effectively running without any AppArmor protection. Use this profile
# only if the process absolutely must be run (effectively) unconfined.
[...]
# Use at your own risk. This profile was developed as an interim workaround for
# LP: #851986 until AppArmor utilizes proper environment filtering.
```
It also contains a line that blocks Python imports (and HexChat plugins use both Perl and Python, which both might require working Chat-Browser IPC) and gives full access to the root directory _inode_, which of course its not required by HexChat:
```
  /opt/google/chrome{,-beta,-unstable}/chrome Pixr,
  /opt/google/chrome{,-beta,-unstable}/{,**/}lib*.so{,.*} m,

  # Full access
  / r,
  /** rwkl,
  /{,usr/,usr/local/}lib{,32,64}/{,**/}*.so{,.*} m,

  # Dangerous files
  audit deny owner /**/* m,              # compiled libraries
  audit deny owner /**/*.py* r,          # python imports
```
Thus, including `abstractions/ubuntu-browsers` to replace a standard `Ux` call might block the existing profiles created for the Firefox variants from working safely outside of the thoroughly preconfigured environment of Ubuntu and may prevent standard Python scripts from working correctly _at least_ when IPC is required.
I truly believe we should prefer the simplest solution, because it guarantees a certain predictability over time and allows the end user easier customization in case of need.

<sub>Edit: clarification of the context, bad copy-paste, grammar, etc..</sub>

**Comment by LorenzoAncora** [comment]
@adrelanos It's been a while and the discussion has gone further than necessary, I hope all the doubts have been clarified. :-)
What have you decided? can we proceed with the merge?

**Comment by adrelanos** [comment]
Probably going to merge this. Just didn't manage yet to test if an
apparmor profile (firefox) will be load either way if it exists.

**Comment by LorenzoAncora** [comment]
> Probably going to merge this. Just didn't manage yet to test if an apparmor profile (firefox) will be load either way if it exists.

@adrelanos ah, is this the doubt that blocks you? You had already asked me to do some tests and I had already verified this situation. Anyway I just checked it out again. ;-)
The tests clearly showed that if the Firefox profile is in enforce mode it is applied in any case: the activation of the link from HexChat only causes the opening of a new tab in the _existing_ process, which if it is already confined remains such; the activation of the link when the browser is _not active_ starts it and is immediately confined in the Firefox profile (the original by Canonical<sup>&reg;</sup>), as shown by the AppArmor output:
```
> sudo aa-status  # AppArmor profiles after activating the link (FF not started)
apparmor module is loaded.
75 profiles are loaded.
56 profiles are in enforce mode.
   [...]
   /usr/bin/xchat
   /usr/lib/cups/backend/cups-pdf
   /usr/lib/firefox{,-[0-9]*}/firefox{,*[^s][^h]}
   [...]
```
AppArmor reports enforce mode for the new Firefox process. In any case, security flaws are mitigated by both the profile and the filtering of the caller's environment. Did I reassure you? ;-)

**Comment by adrelanos** [comment]
This needs to be reverted:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=951331#23

**Pull request merged**
